### DISCUSSION TITLE
ready:script and running:script options?

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#298] [Ideas]
I would like to run a script when the service is started and stopped

Tasks and conditions are perfect for this, however the task only runs once and once it's run successfully it doesn't run again. So it's not ideal for notifications which need to run multiple times.

Do you think it makes sense to have a ready:script (could be paired with the S6 readyness notification) or running:script option? We already have a pre and post option, perhaps these two could be nice compliments?

**Comment by troglobit** [comment]
A ready:script would be really useful I think.  For the mdevd case it's a great fit.

My initial idea was to add a `notify[:condition]` option¹ to the service stanza to unlock the new systemd/s6-like readyness notification mech.  Then the user could use the optional condition arg. to connect that condition to launch a task.  For the mdevd case I think that would've been sufficient, but for the more general case, as you describe above, we need something more flexible and for that a ready:script seems like a good fit.

In summary, we need the new (optional) notification subsystem to integrate better with s6-style daemons like mdevd.  But already now we can add ready:script that would trigger on classic daemons that create their .pid-file when they are done.  The new notification subsystem could then just hook into that same mechanism to trigger the same internal state changes for the service monitor.

______________
¹ much of Finit is built around the condition subsystem, so it would be neat to be able to extend on that.

**Reply by troglobit** [reply]
I've just created issues #299 and #300 for this.

**Reply by troglobit** [reply]
Both tasks now completed and will be included in the upcoming v4.4 release.

**Comment by hongkongkiwi** [comment]
I think there's one usecase I'm missing at the moment, which is being able to retrigger tasks marked as done when the condition changes again. If I want to have a script perform an action when a service runs it will work only the first time, but once the service quits and reruns the task won't trigger again because it is already marked as done.

Is it possible to have something like **alwaysrun:true** so that a task ignores the fact that it has been run successfully ?

This is loosely related to this topic ;) but it made me realise that I'm limited at the moment in some things I want to trigger off these notifications.

**Reply by troglobit** [reply]
I believe a ready:script would achieve that.  But I guess you have other (more interesting) use-cases in mind, perhaps with conditions from multiple sources being asserted to trigger a task more than once?

IIRC we talked about something similar before the summer.  The crux of it is that as long as a condition is asserted, the run/task would be restarted every time it completes.  (Conditions are level triggered).
